base_layer/key_manager/src/cipher_seed.rs

```diff
@@ -264,6 +281,107 @@ impl CipherSeed {
     }
 }
 
+impl CipherSeed {
+    fn generate_hmac(
+        birthday: &[u8],
+        entropy: &[u8],
+        cipher_seed_version: &[u8],
+        salt: &[u8],
+        passphrase: &str,
+    ) -> Result<Vec<u8>, KeyManagerError> {
+        // birthday should be 2 bytes long
+        if birthday.len() != 2 {
+            return Err(KeyManagerError::InvalidData);
+        }
+
+        // entropy should be 16 bytes long
+        if entropy.len() != CIPHER_SEED_ENTROPY_BYTES {
+            return Err(KeyManagerError::InvalidData);
+        }
+
+        // cipher_seed_version should be 1 byte long
+        if cipher_seed_version.len() != 1 {
+            return Err(KeyManagerError::InvalidData);
+        }
+
+        // salt should be 5 bytes long
+        if salt.len() != CIPHER_SEED_SALT_BYTES {
+            return Err(KeyManagerError::InvalidData);
+        }
+
+        // we take the first 32 bytes of the generated derived encryption key for MAC generation, see documentation
+        let passphrase_key: Vec<u8> = Self::generate_domain_separated_passphrase_hash(passphrase, salt)?
+            .into_iter()
+            .rev()
+            .take(32)
+            .rev()
+            .collect();
+
+        // TODO: use hashing API to generate proper domain separation hash
+        let blake2_mac_hasher: VarBlake2b = VarBlake2b::new(CIPHER_SEED_MAC_BYTES).expect(
+            "Should be able to create blake2 hasher; will only panic if output size is 0 or greater than value",
+        );
+        let mut hmac = [0u8; CIPHER_SEED_MAC_BYTES];
+
+        blake2_mac_hasher
+            .chain(DOMAIN_SEPARATION_LABEL.as_bytes())
```

Comment:
As noted elsewhere, ideally this should prepend the length of the label using a fixed-length encoding.

Reply:
I refactored the code to use the hashing API, this safety check should be taken care by the API.